### ISSUE TITLE
Face#svg not documented?

### ISSUE THREAD

**Issue by rtm** [alexbol99/flatten-js#7]
Is it intentional that `Face#svg` is not documented, perhaps because it's intended to be private?

**Comment by alexbol99** [comment]
Yes, my first intention was do not expose `Face#svg` method, but now I see it is worth to do, because sometimes we want to display different faces with different styles. I will add it to the documentation.
Thank you for your remark

**Comment by alexbol99** [comment]
In version v0.5.4 method `face.svg()` now support style parameters.
It is possible now to create svg string where each face has its own style, in the following way:
```
  let svg = ""
  let faces = [...polygon.faces]
  svg += faces[0].svg({stroke:"yellow"})
  svg += faces[1].svg({fill:"none"})
  svg += faces[2].svg({fill:"cyan"}) 
```
And this is documented.

When using in this way, each `face.svg()` call will generate new `<path>` string. This is a default behaviour.
The result will be something like this:
```
<path stroke="yellow" stroke-width="1" fill="lightcyan" fill-opacity="1"   d="
M200,100 L200,300 L440,300 L300,200 L440,150 L500,150 L640,200 L500,300 L740,300 L740,100 L200,100 z" >
</path>
<path stroke="black" stroke-width="1" fill="lightgreen" fill-opacity="1"   d="
M220,150 A30,30 0 0,1 280,150
                    A30,30 0 0,1 220,150 z" >
</path>
<path stroke="black" stroke-width="1" fill="lightblue" fill-opacity="1"   d="
M620,250 A30,30 0 0,1 680,250
                    A30,30 0 0,1 620,250 z" >
</path>
```
Note, that  attribute `fill-rule="evenodd"` has no sense in this case and then this code will not display holes in polygon properly.

The new second parameter `pathDefined=true` enables you to change default behaviour and to call `face.svg()` method inside enclosed user-defined `<path>` string. In this case `face.svg()` method will not create new `<path>` string.

**Comment by alexbol99** [comment]
Issue closed

**Comment by rtm** [comment]
This is poor design (not to mention being incompatible and hence having broken my app, although I know it was never offlicially supported, being undocumented.

Flatten-js is not an SVG library. Even if it were an SVG library, it should not be creating SVG as strings of XML. That is the job for some SVG serialization or de-serialization library, if it's needed at all. SVG is not a string; it's a DOM. Don't build this or any other DOM for me as a string. Depending on how I am using it (for example, I might be using a template-based front-end framework), I don't want, or may not even be able to use, the string-based SVG that you "kindly" created for me. Why should flatten-js know about SVG attributes, or be involved in taking in SVG attributes as object properties and then spitting them right back out again at me as a string? 

In case it's not clear, I want one thing and only one thing from flatten-js as regards SVG, namely a string of the `d` attribute for **me** to use in `path` elements in whatever templating or other environment as I see fit.

To your credit, you anticipated this need, and to address it added the `pathDefined` argument. That's backwards: you changed the default behavior, which required a new parameter to provide the attributes, and then added yet another option to get the previous behavior (which then ignores the first parameter).

It would have been far better to leave `Face#svg` as it was, and just add another interface `Face#svgWithPath(attrs)`.

**Comment by alexbol99** [comment]
Hello @rtm Bob Myers,

For some reason I don't see your message in Guthub, only in e-mail.

First of all, in my optinion there is nothing unusual for abstract geometrical library to implement svg methods. For example, well-known Python Shapely library has them, and they are implemented in the same way: they return XML text string. The reason is that unlike HTML, SVG is not only DOM, but exchangable graphics format. I can create some construction, save it into the file, and then see the file in Adobe Illustrator or any other programm that supports svg graphics. So it is definitely makes sense, especially for testing and quick visualization of simple constructions.

Of couse, this implementation is not suitable for rich interactive graphic application, and it is not supposed to be so.

As for our Face#svg method, I tend to agree with you. I also was not happy with that change, so I will take it back and update documentation. I will inform you about this change in order your application will not be broken again.

By the way, will be nice if you can share with me what kind of application you are working on. It will help me to give better support to you and other users.

You can use my e-mail aexbol99@gmail.com

Best regards,
Alex Bol

**Comment by alexbol99** [comment]
@rtm, this fix reverted in v0.6.3

**Comment by rtm** [comment]
I deleted my original message because it was ill-considered.

I will stick to my humble opinion that it not the job of a library to build strings of SVG DOM for me. I can do that myself. I may be using a library which doesn't work will strings of SVG DOM. What I want is the `d` attribute to the `path` elements, and I will take it from there. 

If you are going to provide an interface which gives someback the entire string representation of the `path` element, then from a design standpoint, for someone like me who just wants the `d` attribute, don't make call the API and pass it a null second argument and then a magic `true` parameter as the third argument. If necessary, just give me another API called `svgPathD` or whatever.

Reverting was good, I guess, but now I have make sure my app didn't break a second time...  :-(

Anyway, thanks for your good work on this project.

I don't know if I ever mentioned that my use case is an app which is used to capture the shape of a room, as part of an apartment inspection app. Originally, I was using your library mainly to handle calculating the area of an arbitrary polygon, and doing some basic 2D stuff. However, as the app evolved, I ended up needing to do some 3D stuff, like calculating where vectors in 3D space (such as the vector pointing down from the camera to the floor) intersected the floor, and also display a real-time 3D preview of the shape being captured, so I ended up bringing in three.js, which is working quite well for me. I still have flatten.js logic in my app for some of the 2D aspects.
